### PR TITLE
Simplify IndexExpression.FindProperty and MethodCallExpression.FindMethod

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses issue #11301.

CC @VSadov, @stephentoub, @justinvp

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs [hunk 1]
```diff
@@ -272,42 +272,24 @@ private static string GetArgTypesString(Expression[] arguments)
 
         private static PropertyInfo FindProperty(Type type, string propertyName, Expression[] arguments, BindingFlags flags)
         {
-            var props = type.GetProperties(flags).Where(x => x.Name.Equals(propertyName, StringComparison.CurrentCultureIgnoreCase)); ;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: These two lines were apparently put in place as a substitution for the missing `FindMembers`, see http://referencesource.microsoft.com/#System.Core/Microsoft/Scripting/Ast/IndexExpression.cs,a2f0d8a96d3c8f44, leading to a bunch of strange logic with redundant checks and `Map` calls. Given that `FindBestProperty` is only used here, I decided to fuse both methods and simplify logic.

`FindMembers` with `Type.FilterNameIgnoreCase` should have the same behavior as our loop with the string equality check using `StringComparison.OrdinalIgnoreCase`, but also saves an array allocation. We don't need LINQ anymore either.

**Comment by bartdesmet** [comment]
Found another one of these in `MethodCallExpression.FindMethod`. Will update the PR after addressing it as well.

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeExtensions.cs [hunk 2]
```diff
@@ -117,19 +117,5 @@ public static MethodInfo GetAnyStaticMethod(this Type type, string name)
             }
             return null;
         }
-
-        public static MethodInfo[] GetMethodsIgnoreCase(this Type type, BindingFlags flags, string name)
-        {
-            var list = new List<MethodInfo>();
-            foreach (var method in type.GetRuntimeMethods())
-            {
-                // TODO: Binding flags filter
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeExtensions.cs]: This caused a regression compared the .NET Framework, changing the reported exception type in a test. See note and change in `ListInitExpressionTests.cs`.

File: src/System.Linq.Expressions/tests/ListInit/ListInitExpressionTests.cs [hunk 3]
```diff
@@ -119,7 +119,10 @@ public void StaticAddMethodOnType()
         {
             var newExp = Expression.New(typeof(EnumerableStaticAdd));
             var adder = typeof(EnumerableStaticAdd).GetMethod(nameof(EnumerableStaticAdd.Add));
-            Assert.Throws<ArgumentException>("addMethod", () => Expression.ListInit(newExp, Expression.Constant("")));
+
+            // this exception behavior (rather than ArgumentException) is compatible with the .NET Framework
+            Assert.Throws<InvalidOperationException>(() => Expression.ListInit(newExp, Expression.Constant("")));
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/tests/ListInit/ListInitExpressionTests.cs]: Repro:
```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

class LI
{
    static void Main()
    {
        var newExp = Expression.New(typeof(EnumerableStaticAdd));
        var adder = typeof(EnumerableStaticAdd).GetMethod(nameof(EnumerableStaticAdd.Add));
        Try(() => Expression.ListInit(newExp, Expression.Constant("")));
        Try(() => Expression.ListInit(newExp, adder, Expression.Constant("")));
        Try(() => Expression.ElementInit(adder, Expression.Constant("")));
        Try(() => Expression.ElementInit(adder, Enumerable.Repeat(Expression.Constant(""), 1)));
    }

    static void Try(Action a)
    {
        try
        {
            a();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.GetType().Name);
        }
    }

    class EnumerableStaticAdd : IEnumerable<string>
    {
        public IEnumerator<string> GetEnumerator() => null;
        IEnumerator IEnumerable.GetEnumerator()  => GetEnumerator();
        public static void Add(string value) { }
    }
}
```
prints (on .NET 4.6)
```csharp
InvalidOperationException
ArgumentException
ArgumentException
ArgumentException
```
We're now compatible.

**Comment by VSadov** [comment]
LGTM. Thanks!!

Looks like the change partially addresses https://github.com/dotnet/corefx/issues/8576

**Pull request merged**